File: 0148-sort-list/0148-sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode merge(ListNode left, ListNode right) {
        if (left == null) return right;
        if (right == null) return left;

        if (left.val < right.val) {
            left.next = merge(left.next, right);
            return left;
        } else {
            right.next = merge(left, right.next);
            return right;
        }
    }

    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) return head;

        ListNode slow = head, fast = head.next;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode mid = slow.next;
        slow.next = null;

        ListNode left = sortList(head);
        ListNode right = sortList(mid);

        return merge(left, right);
    }
}

File: 0148-sort-list/0148-sort-list.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    private fun merge(left: ListNode?, right: ListNode?): ListNode? {
        left ?: return right
        right ?: return left

        return when {
            left.`val` < right.`val` -> left.apply {
                next = merge(next, right)
            }
            else -> right.apply {
                next = merge(left, next)
            }
        }
    }

    fun sortList(head: ListNode?): ListNode? {
        head?.next ?: return head

        var slow = head
        var fast = head.next

        while (fast?.next != null) {
            slow = slow?.next
            fast = fast.next?.next
        }

        val mid = slow?.next
        slow?.next = null

        return merge(sortList(head), sortList(mid))
    }
}